iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,490 @@
+/// <param name="componentName">The component name this property belongs to.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdatePropertyAsync(string propertyName, object propertyValue, PropertyConvention propertyConvention, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Update a collection of properties.
+/// </summary>
+/// <param name="properties">Reported properties to push.</param>
+/// <param name="propertyConvention">A convention handler that defines serializer to use for the properties.</param>
+/// <param name="componentName">The component name this property belongs to.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdatePropertiesAsync(IDictionary<string, object> properties, PropertyConvention propertyConvention, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Update a writable property.
+/// </summary>
+/// <param name="propertyName">Property name.</param>
+/// <param name="writablePropertyResponse">The writable properyt response to push.</param>
+/// <param name="componentName">The component name this property belongs to.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdateWritablePropertyAsync(string propertyName, WritablePropertyResponse writablePropertyResponse, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Sets the global listener for Writable properties
+/// </summary>
+/// <param name="callback">The global call back to handle all writable property updates.</param>
+/// <param name="userContext">Generic parameter to be interpreted by the client code.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task SubscribeToWritablePropertyEvent(Action<Properties, object> callback, object userContext, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Sets the component-level listener for Writable properties
+/// </summary>
+/// <param name="callback">The call back to handle all writable property updates for a particular component.</param>
+/// <param name="componentName">The component name this writable property belongs to.</param>
+/// <param name="userContext">Generic parameter to be interpreted by the client code.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task SubscribeToWritablePropertyEvent(Action<Properties, string, object> callback, string componentName, object userContext, CancellationToken cancellationToken = default);
+```
+
+<details>
+<summary>Properties</summary>
+
+```csharp
+/// <summary>
+/// A container for properties.
+/// </summary>
+public class Properties
+{
+    private PropertyCollection _readOnlyProperties;
```

Comment:
even though Properties contains all readonly properties as a root level collection, it isn't inherently enumerable right now. This is a TODO